File: services/ai/src/tracing.py
import logging
import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.grpc import GrpcAioInstrumentorServer
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

logger = logging.getLogger(__name__)


def setup_tracing() -> None:
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        logger.info("tracing disabled: OTEL_EXPORTER_OTLP_ENDPOINT not set")
        return

    provider = TracerProvider(
        resource=Resource.create(
            {"service.name": os.getenv("OTEL_SERVICE_NAME", "ai-service")}
        )
    )
    provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
    trace.set_tracer_provider(provider)

    GrpcAioInstrumentorServer().instrument()
    HTTPXClientInstrumentor().instrument()
    logger.info("tracing enabled: exporting to %s", endpoint)


def get_span_ids() -> tuple[str, str] | None:
    """Return (trace_id, span_id) of the current span in hex, or None."""
    span = trace.get_current_span()
    context = span.get_span_context()
    if not context.is_valid:
        return None
    return format(context.trace_id, "032x"), format(context.span_id, "016x")

File: services/user/src/observability/tracing.ts
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { IORedisInstrumentation } from '@opentelemetry/instrumentation-ioredis';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { env } from '../config/env.js';
import { logger } from './logger.js';

export function setupTracing(): void {
  if (!env.OTEL_EXPORTER_OTLP_ENDPOINT) {
    logger.info('tracing disabled: OTEL_EXPORTER_OTLP_ENDPOINT not set');
    return;
  }

  const exporter = new OTLPTraceExporter({
    url: new URL('/v1/traces', env.OTEL_EXPORTER_OTLP_ENDPOINT).toString(),
  });

  const sdk = new NodeSDK({
    serviceName: env.OTEL_SERVICE_NAME,
    spanProcessors: [new BatchSpanProcessor(exporter)],
    instrumentations: [new HttpInstrumentation(), new IORedisInstrumentation()],
  });

  sdk.start();
  logger.info('tracing enabled: exporting to %s', env.OTEL_EXPORTER_OTLP_ENDPOINT);
}
